lib/vm/src/mmap.rs

```diff
@@ -256,6 +314,185 @@ impl Mmap {
+                    None => self.len,
+                };
+
+                // The shallow copy failed so we have to do it the hard way
+                let mut off_in: libc::off_t = 0;
+                let mut off_out: libc::off_t = 0;
+
+                cfg_if::cfg_if! {
+                    if #[cfg(not(any(target_env = "musl", target_vendor = "apple")))]
+                    {
+                        let ret = libc::copy_file_range(self.fd.0, &mut off_in, fd.0, &mut off_out, len, 0);
+                    } else {
+                        // TODO: don't use as casts...
+                        let ret = match copy_file_range_impl(self.fd.0, off_in as u64, fd.0, off_out as u64, len) {
+                            Ok(_) => 0,
+                            Err(_err) => -1,
+                        };
+                    }
+                }
+
+                if ret < 0 {
+                    return Err(format!(
+                        "failed to copy temporary file data - {}",
+                        io::Error::last_os_error()
+                    ));
+                }
+
+                #[cfg(feature = "tracing")]
+                trace!("memory copy finished (size={})", len);
+            }
+        }
+
+        // Compute the flags
+        let flags = libc::MAP_FILE | libc::MAP_SHARED;
+
+        // Allocate a single read-write region at once.
+        let ptr = unsafe {
+            libc::mmap(
+                ptr::null_mut(),
+                self.len,
+                libc::PROT_READ | libc::PROT_WRITE,
+                flags,
+                fd.0,
+                0,
+            )
+        };
+        if ptr as isize == -1_isize {
+            return Err(io::Error::last_os_error().to_string());
+        }
+
+        Ok(Self {
+            ptr: ptr as usize,
+            len: self.len,
+            fd,
+        })
+    }
+
+    /// Copies the memory to a new swap file (using copy-on-write if available)
+    #[cfg(target_os = "windows")]
+    pub fn fork(&mut self, hint_used: Option<usize>) -> Result<Self, String> {
+        // Create a new memory which we will copy to
+        let new_mmap = Self::with_at_least(self.len)?;
+
+        #[cfg(feature = "tracing")]
+        trace!("memory copy started");
+
+        // Determine host much to copy
+        let len = match hint_used {
+            Some(a) => a,
+            None => self.len,
+        };
+
+        // Copy the data to the new memory
+        let dst = new_mmap.ptr as *mut u8;
+        let src = self.ptr as *const u8;
+        unsafe {
+            std::ptr::copy_nonoverlapping(src, dst, len);
+        }
+
+        #[cfg(feature = "tracing")]
+        trace!("memory copy finished (size={})", len);
+        Ok(new_mmap)
+    }
+}
+
+/// Rust implementation of libc::copy_file_range.
+///
+/// Needed because that function is not available on all platforms.
+// TODO: better implementation! (this is very quick, low effort)
+// TODO: this function needs tests!
```

Comment:
Add tests

Reply:
Has one test now.

Should get more in the future, but it'll do for now.

Especially since we might refactor this quite a bit to remove the file layer dependency.